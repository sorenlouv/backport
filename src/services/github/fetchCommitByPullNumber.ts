import { BackportOptions } from '../../options/options';
import { CommitSelected } from './Commit';
import { getFormattedCommitMessage } from './commitFormatters';
import { gqlRequest } from './gqlRequest';
import { HandledError } from '../HandledError';

export async function fetchCommitByPullNumber(
  options: BackportOptions & { pullNumber: number }
): Promise<CommitSelected> {
  const {
    githubApiBaseUrlV4,
    repoName,
    repoOwner,
    pullNumber,
    accessToken,
  } = options;
  const query = /* GraphQL */ `
    query getCommitbyPullNumber(
      $repoOwner: String!
      $repoName: String!
      $pullNumber: Int!
    ) {
      repository(owner: $repoOwner, name: $repoName) {
        pullRequest(number: $pullNumber) {
          baseRef {
            name
          }
          mergeCommit {
            oid
            message
          }
        }
      }
    }
  `;

  const res = await gqlRequest<DataResponse>({
    githubApiBaseUrlV4,
    accessToken,
    query,
    variables: {
      repoOwner,
      repoName,
      pullNumber,
    },
  });

  if (res.repository.pullRequest.mergeCommit === null) {
    throw new HandledError(`The PR #${pullNumber} is not merged`);
  }

  const baseBranch = res.repository.pullRequest.baseRef.name;
  const sha = res.repository.pullRequest.mergeCommit.oid;
  const message = getFormattedCommitMessage({
    message: res.repository.pullRequest.mergeCommit.message,
    sha,
    pullNumber,
  });

  return {
    branch: baseBranch,
    sha,
    message,
    pullNumber,
  };
}

interface DataResponse {
  repository: {
    pullRequest: {
      baseRef: {
        name: string;
      };
      mergeCommit: {
        oid: string;
        message: string;
      } | null;
    };
  };
}
