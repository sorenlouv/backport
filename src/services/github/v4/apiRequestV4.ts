import axios from 'axios';
import { AxiosError } from 'axios';
import { HandledError } from '../../HandledError';
import { logger } from '../../logger';

export interface GithubV4Response<DataResponse> {
  data: DataResponse;
  errors?: {
    type: string;
    path: string[];
    locations: {
      line: number;
      column: number;
    }[];
    message: string;
  }[];
}

export async function apiRequestV4<DataResponse>({
  githubApiBaseUrlV4,
  accessToken,
  query,
  variables,
  handleError = true,
}: {
  githubApiBaseUrlV4: string;
  accessToken: string;
  query: string;
  variables?: {
    [key: string]: string | number | null;
  };
  handleError?: boolean;
}) {
  try {
    const response = await axios.post<GithubV4Response<DataResponse>>(
      githubApiBaseUrlV4,
      { query, variables },
      {
        headers: {
          'Content-Type': 'application/json',
          Authorization: `bearer ${accessToken}`,
        },
      }
    );

    if (response.data.errors) {
      const newError = new Error();
      //@ts-ignore
      newError.response = response;
      throw newError;
    }

    logger.info(`POST ${githubApiBaseUrlV4} (status: ${response.status})`);
    logger.verbose('Query:', query);
    logger.verbose('Variables:', variables);
    logger.debug('Response headers:', response.headers);
    logger.verbose('Response data:', response.data);

    return response.data.data;
  } catch (e) {
    logger.info(`POST ${githubApiBaseUrlV4} (status: ${e.response?.status})`);
    logger.info('Query:', query);
    logger.info('Variables:', variables);
    logger.debug('Response headers:', e.response?.headers);
    logger.info('Response data:', e.response?.data);

    if (handleError) {
      throw handleGithubV4Error(e);
    }
    throw e;
  }
}

export function handleGithubV4Error(e: AxiosError<GithubV4Response<null>>) {
  // not github api error
  if (!e.response?.data) {
    return e;
  }

  const errorMessages = e.response.data.errors?.map((error) => error.message);
  if (errorMessages) {
    return new HandledError(`${errorMessages.join(', ')} (Github v4)`);
  }

  return new HandledError(
    `Unexpected response from Github API (v4):\n${JSON.stringify(
      e.response.data,
      null,
      2
    )}`
  );
}
