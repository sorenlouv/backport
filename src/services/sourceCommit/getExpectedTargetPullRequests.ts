import { uniq } from 'lodash';
import { ValidConfigOptions } from '../../options/options';
import { filterNil } from '../../utils/filterEmpty';
import { getFirstLine } from '../github/commitFormatters';
import { parseRemoteConfig } from '../remoteConfig';
import {
  SourcePullRequestNode,
  SourceCommitWithTargetPullRequest,
  TimelineEdge,
  TimelinePullRequestEdge,
} from './parseSourceCommit';

export type ExpectedTargetPullRequest = {
  url?: string;
  number?: number;
  branch: string;
  state: 'OPEN' | 'CLOSED' | 'MERGED' | 'NOT_CREATED';
  mergeCommit?: {
    sha: string;
    message: string;
  };
};

export function getExpectedTargetPullRequests({
  sourceCommit,
  latestBranchLabelMapping,
}: {
  sourceCommit: SourceCommitWithTargetPullRequest;
  latestBranchLabelMapping: ValidConfigOptions['branchLabelMapping'];
}): ExpectedTargetPullRequest[] {
  const sourcePullRequest =
    sourceCommit.associatedPullRequests.edges?.[0]?.node;

  const remoteConfig =
    sourcePullRequest?.mergeCommit.remoteConfigHistory.edges?.[0]?.remoteConfig;

  const branchLabelMapping =
    (remoteConfig && parseRemoteConfig(remoteConfig)?.branchLabelMapping) ??
    latestBranchLabelMapping;

  // if there is no source pull request the commit was pushed directly to the source branch
  // in that case there will be no labels, and thus not possible to deduce the expected target branches
  if (!sourcePullRequest) {
    return [];
  }

  const existingTargetPullRequests = getExistingTargetPullRequests(
    sourceCommit,
    sourcePullRequest
  );

  // if there's no `branchLabelMapping`, it's not possible to deduce the missing target branches
  if (!branchLabelMapping) {
    return existingTargetPullRequests;
  }

  const missingTargetPullRequests = getMissingTargetPullRequests(
    sourcePullRequest,
    existingTargetPullRequests,
    branchLabelMapping
  );

  return [...existingTargetPullRequests, ...missingTargetPullRequests];
}

function getExistingTargetPullRequests(
  sourceCommit: SourceCommitWithTargetPullRequest,
  sourcePullRequest: SourcePullRequestNode
): ExpectedTargetPullRequest[] {
  const sourceCommitMessage = getFirstLine(sourceCommit.message);

  return sourcePullRequest.timelineItems.edges
    .filter(filterNil)
    .filter(filterPullRequests)
    .filter((item) => {
      const { targetPullRequest } = item.node;

      // ignore closed PRs
      if (targetPullRequest.state === 'CLOSED') {
        return false;
      }

      // at least one of the commits in `targetPullRequest` should match the merge commit from the source pull request
      const didCommitMatch = targetPullRequest.commits.edges.some(
        (commitEdge) => {
          const { targetCommit } = commitEdge.node;

          const matchingRepoName =
            sourceCommit.repository.name === targetPullRequest.repository.name;

          const matchingRepoOwner =
            sourceCommit.repository.owner.login ===
            targetPullRequest.repository.owner.login;

          const targetCommitMessage = getFirstLine(targetCommit.message);

          const matchingMessage = targetCommitMessage === sourceCommitMessage;
          return matchingRepoName && matchingRepoOwner && matchingMessage;
        }
      );

      const titleIncludesMessage =
        targetPullRequest.title.includes(sourceCommitMessage);

      const titleIncludesNumber = targetPullRequest.title.includes(
        sourcePullRequest.number.toString()
      );

      return didCommitMatch || (titleIncludesMessage && titleIncludesNumber);
    })
    .map((item) => {
      const { targetPullRequest } = item.node;
      return {
        url: targetPullRequest.url,
        number: targetPullRequest.number,
        branch: targetPullRequest.baseRefName,
        state: targetPullRequest.state,
        mergeCommit: targetPullRequest.targetMergeCommit
          ? {
              sha: targetPullRequest.targetMergeCommit.sha,
              message: targetPullRequest.targetMergeCommit.message,
            }
          : undefined,
      };
    });
}

function getMissingTargetPullRequests(
  sourcePullRequest: SourcePullRequestNode,
  existingTargetPullRequests: ExpectedTargetPullRequest[],
  branchLabelMapping: NonNullable<ValidConfigOptions['branchLabelMapping']>
): ExpectedTargetPullRequest[] {
  const labels = sourcePullRequest.labels.nodes.map((label) => label.name);
  const targetBranchesFromLabels = labels
    .map((label) => getTargetBranchForLabel({ branchLabelMapping, label }))
    .filter(filterNil)
    .filter((targetBranch) => targetBranch !== sourcePullRequest.baseRefName);

  const expectedTargetBranches = existingTargetPullRequests.map(
    (pr) => pr.branch
  );
  const expected = uniq(targetBranchesFromLabels);
  return expected
    .filter((targetBranch) => !expectedTargetBranches.includes(targetBranch))
    .map((branch) => {
      return { branch, state: 'NOT_CREATED' as const };
    });
}

// narrow TimelineEdge to TimelinePullRequestEdge
function filterPullRequests(
  item: TimelineEdge
): item is TimelinePullRequestEdge {
  const { targetPullRequest } = item.node;
  return targetPullRequest.__typename === 'PullRequest';
}

function getTargetBranchForLabel({
  branchLabelMapping,
  label,
}: {
  branchLabelMapping: NonNullable<ValidConfigOptions['branchLabelMapping']>;
  label: string;
}) {
  // only get first match
  const result = Object.entries(branchLabelMapping).find(([labelPattern]) => {
    const regex = new RegExp(labelPattern);
    const isMatch = label.match(regex) !== null;
    return isMatch;
  });

  if (result) {
    const [labelPattern, targetBranchPattern] = result;
    const regex = new RegExp(labelPattern);
    const targetBranch = label.replace(regex, targetBranchPattern);
    if (targetBranch) {
      return targetBranch;
    }
  }
}
