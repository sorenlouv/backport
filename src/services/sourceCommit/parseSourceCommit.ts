import { isEmpty } from 'lodash';
import { ValidConfigOptions } from '../../options/options';
import { extractPullNumber } from '../github/commitFormatters';
import {
  ExpectedTargetPullRequest,
  getExpectedTargetPullRequests,
} from './getExpectedTargetPullRequests';

export interface Commit {
  committedDate: string;
  sourceBranch: string;
  sha: string;
  originalMessage: string; // TODO: rename to message
  pullNumber?: number;
  pullUrl?: string;
  expectedTargetPullRequests: ExpectedTargetPullRequest[];
}

export interface PullRequestNode {
  baseRefName: string;
  url: string;
  number: number;
  labels: {
    nodes: {
      name: string;
    }[];
  };
  timelineItems: {
    edges: Array<TimelineEdge>;
  };
}

export type TimelineEdge = TimelinePullRequestEdge | TimelineIssueEdge;

export interface TimelinePullRequestEdge {
  node: {
    targetPullRequest: {
      __typename: 'PullRequest';
      url: string;
      title: string;
      state: 'OPEN' | 'CLOSED' | 'MERGED';
      baseRefName: string;
      number: number;
      commits: {
        edges: Array<{
          node: { targetCommit: InnerCommitNode };
        }>;
      };
    };
  };
}

interface TimelineIssueEdge {
  node: { targetPullRequest: { __typename: 'Issue' } };
}

type InnerCommitNode = {
  repository: {
    name: string;
    owner: { login: string };
  };
  oid: string;
  message: string;
  committedDate: string;
};

export type SourceCommitWithTargetPullRequest = InnerCommitNode & {
  associatedPullRequests: {
    edges: { node: PullRequestNode }[] | null;
  };
};

export function parseSourceCommit({
  sourceCommit,
  options,
}: {
  sourceCommit: SourceCommitWithTargetPullRequest;
  options: ValidConfigOptions;
}): Commit {
  const sourcePullRequest =
    sourceCommit.associatedPullRequests.edges?.[0]?.node;
  const commitMessage = sourceCommit.message;
  const pullNumber =
    sourcePullRequest?.number ?? extractPullNumber(commitMessage);
  const sourceBranch = sourcePullRequest?.baseRefName ?? options.sourceBranch;
  const branchLabelMapping = getBranchLabelMappingForCommit(
    sourceCommit,
    options.branchLabelMapping,
    options.historicalBranchLabelMappings
  );

  const expectedTargetPullRequests = getExpectedTargetPullRequests(
    sourceCommit,
    branchLabelMapping
  );

  return {
    committedDate: sourceCommit.committedDate,
    sourceBranch,
    sha: sourceCommit.oid,
    originalMessage: commitMessage,
    pullNumber,
    pullUrl: sourcePullRequest?.url,
    expectedTargetPullRequests,
  };
}

export const sourceCommitWithTargetPullRequestFragment = {
  name: 'SourceCommitWithTargetPullRequest',
  source: /* GraphQL */ `
    fragment InnerCommitNode on Commit {
      repository {
        name
        owner {
          login
        }
      }
      oid
      message
      committedDate
    }

    fragment SourceCommitWithTargetPullRequest on Commit {
      ...InnerCommitNode
      associatedPullRequests(first: 1) {
        edges {
          node {
            # Source PR
            url
            number
            labels(first: 50) {
              nodes {
                name
              }
            }
            baseRefName
            timelineItems(last: 20, itemTypes: CROSS_REFERENCED_EVENT) {
              edges {
                node {
                  ... on CrossReferencedEvent {
                    targetPullRequest: source {
                      __typename

                      # Target PRs
                      ... on PullRequest {
                        url
                        title
                        state
                        baseRefName
                        number
                        commits(first: 20) {
                          edges {
                            node {
                              targetCommit: commit {
                                ...InnerCommitNode
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  `,
};

function getBranchLabelMappingForCommit(
  sourceCommit: SourceCommitWithTargetPullRequest,
  branchLabelMapping: ValidConfigOptions['branchLabelMapping'],
  historicalBranchLabelMappings: ValidConfigOptions['historicalBranchLabelMappings']
): Record<string, string> | undefined {
  if (isEmpty(historicalBranchLabelMappings)) {
    return branchLabelMapping;
  }

  const match = historicalBranchLabelMappings.find(
    (branchLabelMapping) =>
      sourceCommit.committedDate > branchLabelMapping.committedDate
  );

  return (
    match?.branchLabelMapping ??
    historicalBranchLabelMappings[0]?.branchLabelMapping
  );
}
