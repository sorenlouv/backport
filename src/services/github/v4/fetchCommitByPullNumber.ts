import chalk from 'chalk';
import ora from 'ora';
import { ValidConfigOptions } from '../../../options/options';
import { Commit } from '../../../types/Commit';
import { HandledError } from '../../HandledError';
import { getFormattedCommitMessage } from '../commitFormatters';
import { apiRequestV4 } from './apiRequestV4';
import {
  PullRequestNode,
  pullRequestFragment,
  pullRequestFragmentName,
  getExistingTargetPullRequests,
  getPullRequestLabels,
} from './getExistingTargetPullRequests';
import { getTargetBranchesFromLabels } from './getTargetBranchesFromLabels';

export async function fetchCommitByPullNumber(
  options: ValidConfigOptions & { pullNumber: number }
): Promise<Commit> {
  const {
    accessToken,
    githubApiBaseUrlV4,
    pullNumber,
    repoName,
    repoOwner,
  } = options;
  const query = /* GraphQL */ `
    query CommitByPullNumber(
      $repoOwner: String!
      $repoName: String!
      $pullNumber: Int!
    ) {
      repository(owner: $repoOwner, name: $repoName) {
        pullRequest(number: $pullNumber) {
          ...${pullRequestFragmentName}
        }
      }
    }

    ${pullRequestFragment}
  `;

  const spinner = ora(
    `Loading merge commit from pull request #${options.pullNumber}`
  ).start();

  let res: DataResponse;
  try {
    res = await apiRequestV4<DataResponse>({
      githubApiBaseUrlV4,
      accessToken,
      query,
      variables: {
        repoOwner,
        repoName,
        pullNumber,
      },
    });
    spinner.stop();
  } catch (e) {
    spinner.fail();
    throw e;
  }

  const pullRequestNode = res.repository.pullRequest;

  if (pullRequestNode.mergeCommit === null) {
    throw new HandledError(`The PR #${pullNumber} is not merged`);
  }

  const sourceBranch = pullRequestNode.baseRefName;
  const sha = pullRequestNode.mergeCommit.oid;
  const commitMessage = pullRequestNode.mergeCommit.message;
  const formattedMessage = getFormattedCommitMessage({
    message: commitMessage,
    sha,
    pullNumber,
  });

  // add styles to make it look like a prompt question
  spinner.stopAndPersist({
    symbol: chalk.green('?'),
    text: `${chalk.bold('Select pull request')} ${chalk.cyan(
      formattedMessage
    )}`,
  });

  const existingTargetPullRequests = getExistingTargetPullRequests(
    pullRequestNode
  );

  const targetBranchesFromLabels = getTargetBranchesFromLabels({
    sourceBranch: pullRequestNode.baseRefName,
    existingTargetPullRequests,
    branchLabelMapping: options.branchLabelMapping,
    labels: getPullRequestLabels(pullRequestNode),
  });

  return {
    sourceBranch,
    targetBranchesFromLabels,
    sha,
    formattedMessage,
    originalMessage: commitMessage,
    pullNumber,
    existingTargetPullRequests,
  };
}

interface DataResponse {
  repository: {
    pullRequest: PullRequestNode;
  };
}
