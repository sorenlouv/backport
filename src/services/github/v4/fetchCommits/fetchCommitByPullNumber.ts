import { ValidConfigOptions } from '../../../../options/options';
import { HandledError } from '../../../HandledError';
import {
  Commit,
  SourceCommitWithTargetPullRequest,
  sourceCommitWithTargetPullRequestFragment,
  parseSourceCommit,
} from '../../../sourceCommit/parseSourceCommit';
import { apiRequestV4 } from '../apiRequestV4';

export async function fetchCommitByPullNumber(options: {
  accessToken: string;
  branchLabelMapping?: ValidConfigOptions['branchLabelMapping'];
  githubApiBaseUrlV4?: string;
  historicalBranchLabelMappings: ValidConfigOptions['historicalBranchLabelMappings'];
  pullNumber: number;
  repoName: string;
  repoOwner: string;
  sourceBranch: string;
}): Promise<Commit> {
  const {
    accessToken,
    githubApiBaseUrlV4 = 'https://api.github.com/graphql',
    pullNumber,
    repoName,
    repoOwner,
  } = options;

  const query = /* GraphQL */ `
    query CommitByPullNumber(
      $repoOwner: String!
      $repoName: String!
      $pullNumber: Int!
    ) {
      repository(owner: $repoOwner, name: $repoName) {
        pullRequest(number: $pullNumber) {
          mergeCommit {
            ...SourceCommitWithTargetPullRequest
          }
        }
      }
    }

    ${sourceCommitWithTargetPullRequestFragment.source}
  `;

  const res = await apiRequestV4<CommitByPullNumberResponse>({
    githubApiBaseUrlV4,
    accessToken,
    query,
    variables: {
      repoOwner,
      repoName,
      pullNumber,
    },
  });

  const pullRequestNode = res.repository.pullRequest;
  if (!pullRequestNode) {
    throw new HandledError(`The PR #${pullNumber} does not exist`);
  }

  const sourceCommit = pullRequestNode.mergeCommit;
  if (sourceCommit === null) {
    throw new HandledError(`The PR #${pullNumber} is not merged`);
  }
  return parseSourceCommit({ options, sourceCommit });
}

interface CommitByPullNumberResponse {
  repository: {
    pullRequest: {
      mergeCommit: SourceCommitWithTargetPullRequest | null;
    } | null;
  };
}
