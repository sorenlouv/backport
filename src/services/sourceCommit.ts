import { ValidConfigOptions } from '../options/options';
import {
  getFirstCommitMessageLine,
  getFormattedCommitMessage,
  getPullNumberFromMessage,
} from '../services/github/commitFormatters';
import { getTargetBranchesFromLabels } from '../services/github/v4/getTargetBranchesFromLabels';
import { filterNil } from '../utils/filterEmpty';

export interface Commit {
  committedDate: string;
  sourceBranch: string;
  targetBranchesFromLabels: {
    expected: string[];
    missing: string[];
    unmerged: string[];
    merged: string[];
  };
  sha: string;
  formattedMessage: string;
  originalMessage: string;
  pullNumber?: number;
  pullUrl?: string;
  existingTargetPullRequests: ExistingTargetPullRequests;
}

export interface PullRequestNode {
  baseRefName: string;
  url: string;
  number: number;
  labels: {
    nodes: {
      name: string;
    }[];
  };
  timelineItems: {
    edges: Array<TimelineEdge>;
  };
}

type TimelineEdge = TimelinePullRequestEdge | TimelineIssueEdge;

interface TimelinePullRequestEdge {
  node: {
    targetPullRequest: {
      __typename: 'PullRequest';
      url: string;
      title: string;
      state: 'OPEN' | 'CLOSED' | 'MERGED';
      baseRefName: string;
      number: number;
      commits: {
        edges: Array<{
          node: { targetCommit: InnerCommitNode };
        }>;
      };
    };
  };
}

interface TimelineIssueEdge {
  node: { targetPullRequest: { __typename: 'Issue' } };
}

type InnerCommitNode = {
  repository: {
    name: string;
    owner: { login: string };
  };
  oid: string;
  message: string;
  committedDate: string;
};

export type SourceCommitWithTargetPullRequest = InnerCommitNode & {
  associatedPullRequests: {
    edges: { node: PullRequestNode }[] | null;
  };
};

function getPullRequestLabels(pullRequestNode?: PullRequestNode) {
  return pullRequestNode?.labels.nodes.map((label) => label.name);
}

export function parseSourceCommit({
  sourceCommit,
  options,
}: {
  sourceCommit: SourceCommitWithTargetPullRequest;
  options: ValidConfigOptions;
}): Commit {
  const existingTargetPullRequests =
    getExistingTargetPullRequests(sourceCommit);
  const pullRequestNode = sourceCommit.associatedPullRequests.edges?.[0]?.node;
  const commitMessage = sourceCommit.message;
  const sha = sourceCommit.oid;
  const pullNumber =
    pullRequestNode?.number ?? getPullNumberFromMessage(commitMessage);

  const sourceBranch = pullRequestNode?.baseRefName ?? options.sourceBranch;
  const targetBranchesFromLabels = getTargetBranchesFromLabels({
    sourceBranch,
    existingTargetPullRequests,
    branchLabelMapping: options.branchLabelMapping,
    labels: getPullRequestLabels(pullRequestNode),
  });

  const formattedMessage = getFormattedCommitMessage({
    message: commitMessage,
    pullNumber,
    sha,
  });

  return {
    committedDate: sourceCommit.committedDate,
    sourceBranch, // TODO can this be deleted since it's already in `options`?
    targetBranchesFromLabels,
    sha: sourceCommit.oid,
    formattedMessage,
    originalMessage: commitMessage,
    pullNumber,
    pullUrl: pullRequestNode?.url,
    existingTargetPullRequests,
  };
}

export const sourceCommitWithTargetPullRequestFragment = {
  name: 'SourceCommitWithTargetPullRequest',
  source: /* GraphQL */ `
    fragment InnerCommitNode on Commit {
      repository {
        name
        owner {
          login
        }
      }
      oid
      message
      committedDate
    }

    fragment SourceCommitWithTargetPullRequest on Commit {
      ...InnerCommitNode
      associatedPullRequests(first: 1) {
        edges {
          node {
            # Source PR
            url
            number
            labels(first: 50) {
              nodes {
                name
              }
            }
            baseRefName
            timelineItems(last: 20, itemTypes: CROSS_REFERENCED_EVENT) {
              edges {
                node {
                  ... on CrossReferencedEvent {
                    targetPullRequest: source {
                      __typename

                      # Target PRs
                      ... on PullRequest {
                        url
                        title
                        state
                        baseRefName
                        number
                        commits(first: 20) {
                          edges {
                            node {
                              targetCommit: commit {
                                ...InnerCommitNode
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  `,
};

export type ExistingTargetPullRequests = ReturnType<
  typeof getExistingTargetPullRequests
>;
export function getExistingTargetPullRequests(
  sourceCommit: SourceCommitWithTargetPullRequest
) {
  const sourcePullRequest =
    sourceCommit.associatedPullRequests.edges?.[0]?.node;
  if (!sourcePullRequest) {
    return [];
  }

  const sourceCommitMessage = getFirstCommitMessageLine(sourceCommit.message);

  return sourcePullRequest.timelineItems.edges
    .filter(filterNil)
    .filter(filterPullRequests)
    .filter((item) => {
      const { targetPullRequest } = item.node;

      // ignore closed PRs
      if (targetPullRequest.state === 'CLOSED') {
        return false;
      }

      // at least one of the commits in `targetPullRequest` should match the merge commit from the source pull request
      const didCommitMatch = targetPullRequest.commits.edges.some(
        (commitEdge) => {
          const { targetCommit } = commitEdge.node;

          const matchingRepoName =
            sourceCommit.repository.name === targetCommit.repository.name;

          const matchingRepoOwner =
            sourceCommit.repository.owner.login ===
            targetCommit.repository.owner.login;

          const targetCommitMessage = getFirstCommitMessageLine(
            targetCommit.message
          );

          const matchingMessage = targetCommitMessage === sourceCommitMessage;

          return matchingRepoName && matchingRepoOwner && matchingMessage;
        }
      );

      const titleIncludesMessage =
        targetPullRequest.title.includes(sourceCommitMessage);

      const titleIncludesNumber = targetPullRequest.title.includes(
        sourcePullRequest.number.toString()
      );

      return didCommitMatch || (titleIncludesMessage && titleIncludesNumber);
    })
    .map((item) => {
      const { targetPullRequest } = item.node;
      return {
        url: targetPullRequest.url,
        number: targetPullRequest.number,
        branch: targetPullRequest.baseRefName,
        state: targetPullRequest.state,
      };
    });
}

// narrow TimelineEdge to TimelinePullRequestEdge
function filterPullRequests(
  item: TimelineEdge
): item is TimelinePullRequestEdge {
  const { targetPullRequest } = item.node;
  return targetPullRequest.__typename === 'PullRequest';
}
