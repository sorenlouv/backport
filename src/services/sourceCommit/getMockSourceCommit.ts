import { SourceCommitWithTargetPullRequest } from './parseSourceCommit';

export function getMockSourceCommit({
  sourceCommitMessage,
  sourcePullRequest,
  timelineItems = [],
}: {
  sourceCommitMessage: string;
  sourcePullRequest: {
    number: number;
    labels?: string[];
    sourceBranch?: string;
  } | null;
  timelineItems?: Array<{
    state: 'OPEN' | 'CLOSED' | 'MERGED';
    targetBranch: string;
    title?: string;
    number: number;
    commitMessages: string[];
    repoName?: string;
    repoOwner?: string;
  }>;
}) {
  const sourceCommit: SourceCommitWithTargetPullRequest = {
    repository: {
      name: 'kibana',
      owner: { login: 'elastic' },
    },
    committedDate: '2021-12-22T00:00:00Z',
    oid: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
    message: sourceCommitMessage,
    associatedPullRequests: {
      edges: !sourcePullRequest
        ? null
        : [
            {
              node: {
                url: `https://github.com/elastic/kibana/pull/${sourcePullRequest.number}`,
                labels: {
                  nodes: (sourcePullRequest.labels ?? []).map((name) => ({
                    name,
                  })),
                },
                baseRefName:
                  sourcePullRequest.sourceBranch ??
                  'source-branch-from-associated-pull-request',
                number: sourcePullRequest.number,
                timelineItems: {
                  edges: timelineItems.map((timelineItem) => {
                    return {
                      node: {
                        targetPullRequest: {
                          __typename: 'PullRequest',
                          url: `https://github.com/elastic/kibana/pull/${timelineItem.number}`,
                          title:
                            timelineItem.title ?? 'Default PR title (#123)',
                          number: timelineItem.number,
                          state: timelineItem.state,
                          baseRefName: timelineItem.targetBranch,
                          commits: {
                            edges: timelineItem.commitMessages.map(
                              (message) => ({
                                node: {
                                  targetCommit: {
                                    repository: {
                                      name: timelineItem.repoName ?? 'kibana',
                                      owner: {
                                        login:
                                          timelineItem.repoOwner ?? 'elastic',
                                      },
                                    },
                                    committedDate: '2021-12-23T00:00:00Z',
                                    oid: 'abc',
                                    message: message,
                                  },
                                },
                              })
                            ),
                          },
                        },
                      },
                    };
                  }),
                },
              },
            },
          ],
    },
  };
  return sourceCommit;
}
