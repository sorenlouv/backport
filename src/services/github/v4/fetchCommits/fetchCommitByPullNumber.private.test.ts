import { print } from 'graphql';
import { getDevAccessToken } from '../../../../test/private/getDevAccessToken';
import { Commit } from '../../../sourceCommit/parseSourceCommit';
import * as apiRequestV4Module from '../apiRequestV4';
import { fetchCommitByPullNumber } from './fetchCommitByPullNumber';

const accessToken = getDevAccessToken();

describe('fetchCommitByPullNumber', () => {
  describe('snapshot request/response', () => {
    let spy: jest.SpyInstance;
    let commit: Commit;

    beforeEach(async () => {
      spy = jest.spyOn(apiRequestV4Module, 'apiRequestV4');

      commit = await fetchCommitByPullNumber({
        repoOwner: 'elastic',
        repoName: 'kibana',
        accessToken,
        pullNumber: 121633,
        sourceBranch: 'master',
      });
    });

    it('makes the right queries', () => {
      const queries = spy.mock.calls.reduce((acc, call) => {
        const query = call[0].query;
        const name = apiRequestV4Module.getQueryName(query);
        return { ...acc, [name]: print(query) };
      }, {});

      const queryNames = Object.keys(queries);
      expect(queryNames).toEqual(['CommitByPullNumber']);

      queryNames.forEach((name) => {
        expect(queries[name]).toMatchSnapshot(`Query: ${name}`);
      });
    });

    it('returns the correct response', async () => {
      expect(commit).toMatchSnapshot();
    });
  });

  describe('when PR was merged', () => {
    it('is returned', async () => {
      const options = {
        accessToken,
        pullNumber: 5,
        repoName: 'backport-e2e',
        repoOwner: 'backport-org',
        sourceBranch: 'main',
      };

      const expectedCommit: Commit = {
        author: { email: 'sorenlouv@gmail.com', name: 'Søren Louv-Jansen' },
        sourceCommit: {
          committedDate: '2020-08-15T12:40:19Z',
          message: 'Add 🍏 emoji (#5)',
          sha: 'ee8c492334cef1ca077a56addb79a26f79821d2f',
        },
        sourcePullRequest: {
          number: 5,
          url: 'https://github.com/backport-org/backport-e2e/pull/5',
          mergeCommit: {
            message: 'Add 🍏 emoji (#5)',
            sha: 'ee8c492334cef1ca077a56addb79a26f79821d2f',
          },
        },
        sourceBranch: 'master',
        expectedTargetPullRequests: [
          {
            branch: '7.x',
            state: 'MERGED',
            number: 6,
            url: 'https://github.com/backport-org/backport-e2e/pull/6',
            mergeCommit: {
              message: 'Add 🍏 emoji (#5) (#6)',
              sha: '4bcd876d4ceaa73cf437bfc89b74d1a4e704c0a6',
            },
          },
          {
            branch: '7.8',
            state: 'MERGED',
            number: 7,
            url: 'https://github.com/backport-org/backport-e2e/pull/7',
            mergeCommit: {
              message: 'Add 🍏 emoji (#5) (#7)',
              sha: '46cd6f9999effdf894a36dbc7db90e890f4be840',
            },
          },
        ],
      };

      expect(await fetchCommitByPullNumber(options)).toEqual(expectedCommit);
    });
  });

  describe('when PR is still open', () => {
    it('throws an error', async () => {
      const options = {
        accessToken,
        pullNumber: 11,
        repoName: 'backport-e2e',
        repoOwner: 'backport-org',
        sourceBranch: 'main',
      };

      await expect(fetchCommitByPullNumber(options)).rejects.toThrowError(
        `The PR #11 is not merged`
      );
    });
  });

  describe('when PR does not exist', () => {
    it('throws an error', async () => {
      const options = {
        accessToken,
        pullNumber: 9999999999999,
        repoName: 'backport-e2e',
        repoOwner: 'backport-org',
        sourceBranch: 'main',
      };

      await expect(fetchCommitByPullNumber(options)).rejects.toThrowError(
        `Could not resolve to a PullRequest with the number of 9999999999999. (Github API v4)`
      );
    });
  });
});
