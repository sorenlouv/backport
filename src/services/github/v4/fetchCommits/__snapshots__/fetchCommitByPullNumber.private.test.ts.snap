// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchCommitByPullNumber snapshot request/response makes the right queries: CommitsBySha 1`] = `undefined`;

exports[`fetchCommitByPullNumber snapshot request/response returns the correct response 1`] = `
Object {
  "committedDate": "2021-12-20T14:20:16Z",
  "expectedTargetPullRequests": Array [
    Object {
      "branch": "8.0",
      "mergeCommit": Object {
        "message": "[APM] Add note about synthtrace to APM docs (#121633) (#121643)

Co-authored-by: SÃ¸ren Louv-Jansen <soren.louv@elastic.co>",
        "sha": "842adfdeb5541b059231857522f9009771a46107",
      },
      "number": 121643,
      "state": "MERGED",
      "url": "https://github.com/elastic/kibana/pull/121643",
    },
  ],
  "originalMessage": "[APM] Add note about synthtrace to APM docs (#121633)",
  "pullNumber": 121633,
  "pullUrl": "https://github.com/elastic/kibana/pull/121633",
  "sha": "d421ddcf6157150596581c7885afa3690cec6339",
  "sourceBranch": "main",
}
`;
